src/NServiceBus.Core/DeliveryConstraints/DeliveryConstraintContextExtensions.cs

```diff
@@ -3,55 +3,45 @@ namespace NServiceBus.DeliveryConstraints
     using System;
     using System.Collections.Generic;
     using System.Linq;
+    using NServiceBus.Extensibility;
     using NServiceBus.Features;
-    using NServiceBus.Pipeline;
     using NServiceBus.Transports;
 
     /// <summary>
-    /// Gives access to <see cref="DeliveryConstraint"/>s that exist in the various <see cref="IBehaviorContext"/>s.
+    /// Gives access to <see cref="DeliveryConstraint"/>s that exist in the various <see cref="ContextBag"/>s.
     /// </summary>
     public static class DeliveryConstraintContextExtensions
     {
         /// <summary>
-        /// Adds a <see cref="DeliveryConstraint"/> to a <see cref="IRoutingContext"/>.
+        /// Adds a <see cref="DeliveryConstraint"/> to a <see cref="ContextBag"/>.
         /// </summary>
-        public static void AddDeliveryConstraint(this IRoutingContext context, DeliveryConstraint constraint)
+        public static void AddDeliveryConstraint(this ContextBag context, DeliveryConstraint constraint)
         {
-            AddDeliveryConstraintInternal(context, constraint);
-        }
-
-        /// <summary>
-        /// Adds a <see cref="DeliveryConstraint"/> to a <see cref="IOutgoingLogicalMessageContext"/>.
-        /// </summary>
-        public static void AddDeliveryConstraint(this IOutgoingLogicalMessageContext context, DeliveryConstraint constraint)
-        {
-            AddDeliveryConstraintInternal(context, constraint);
-        }
+            List<DeliveryConstraint> constraints;
 
+            if (!context.TryGet(out constraints))
+            {
+                constraints = new List<DeliveryConstraint>();
 
-        /// <summary>
-        /// Tries to retrieves an instance of <typeparamref name="T"/> from a <see cref="IRoutingContext"/>.
-        /// </summary>
-        public static bool TryGetDeliveryConstraint<T>(this IRoutingContext context, out T constraint) where T : DeliveryConstraint
-        {
-            List<DeliveryConstraint> constraints;
+                context.Set(constraints);
+            }
 
-            if (context.Extensions.TryGet(out constraints))
+            if (constraints.Any(c => c.GetType() == constraint.GetType()))
```

Comment:
Maybe this is neglectable because we should not have many delivery constraints. But still I want to mention it here:

I think we should try to avoid LINQ if we can in operations inside the pipeline for performance and mem allocation reasons. Should we just use a Set or a Dictionary here?